src/sequential/sequential.cpp

```diff
@@ -0,0 +1,275 @@
+    const auto [status, offset] = get_wouldbe_offset(tmp_buckets, tmp.key, tmp.hashcode, home);
+    switch (status) {
+      case FOUND_MATCH: {
+        LOG_DEBUG("FOUND_MATCH");
+        size_t real_index = get_probe(home, offset, tmp_buckets.size());
+        SequentialBucket &bkt = tmp_buckets[real_index];
+        bkt.value = tmp.value;
+        return 0;
+      }
+      case FOUND_SWAP: {
+        LOG_DEBUG("FOUND_SWAP");
+        size_t real_index = get_probe(home, offset, tmp_buckets.size());
+        SequentialBucket &bkt = tmp_buckets[real_index];
+        tmp.offset = offset;
+        bkt.swap(tmp);
+        // TODO(dchu): likely to be buggy
+        continue;
+      }
+      case FOUND_HOLE: {
+        LOG_DEBUG("FOUND_HOLE");
+        size_t real_index = get_probe(home, offset, tmp_buckets.size());
+        SequentialBucket &bkt = tmp_buckets[real_index];
+        tmp.offset = offset;
+        bkt.swap(tmp);
+        // TODO(dchu): likely to be buggy
+        return 0;
+      }
+      case FOUND_NOHOLE:
+        assert(0 && "no hole found!");
+        exit(1);
+      default:
+        assert(0 && "impossible!");
+        exit(1);
+    }
+  }
+  assert(0 && "impossible!");
+  exit(1);
+}
+
+
+////////////////////////////////////////////////////////////////////////////////
+/// HASH TABLE CLASS
+////////////////////////////////////////////////////////////////////////////////
+
+
+
+int SequentialRobinHoodHashTable::insert(KeyType key, ValueType value) {
+  LOG_TRACE("Enter");
+  // TODO: ensure key and value are valid
+  // 1. Error check arguments
+  // 2. Check if already present
+  // 3.   If not, check if room to insert
+  // 4.     If not, resize
+  // 5. Insert (with swapping if necessary)
+  int r = 0;
+  HashCodeType hashcode = hash(key);
+  size_t home = get_home(hashcode, this->size);
+
+  const auto [status, offset] = get_wouldbe_offset(this->buckets, key, hashcode, home);
+  switch (status) {
+  case FOUND_MATCH:
+    LOG_DEBUG("FOUND_MATCH");
+    r = insert_without_resize(this->buckets, key, value, hashcode);
+    assert(!r && "error in insert_without_resize");
+    return r;
+  case FOUND_HOLE:
+    LOG_DEBUG("FOUND_HOLE");
+    r = insert_without_resize(this->buckets, key, value, hashcode);
+    assert(!r && "error in insert_without_resize");
+    ++this->length;
+    return r;
+  case FOUND_SWAP:
+  case FOUND_NOHOLE:
+    LOG_DEBUG("FOUND_{SWAP,NOHOLE}");
+    // Ensure suitably empty and there is at least one hole
+    if (this->length >= 0.9 * this->size || this->length + 1 >= this->size) {
+      int r = this->resize(2 * this->size);
+      assert(!r && "error in resize");
+    }
+    r = insert_without_resize(this->buckets, key, value, hashcode);
+    assert(!r && "error in insert_without_resize");
+    ++this->length;
+    return r;
+  default:
+    assert(0 && "impossible");
+  }
+  assert(0 && "unreachable");
+}
+
+int SequentialRobinHoodHashTable::search(KeyType key, ValueType *value) {
```

Comment:
```suggestion
std::optional<ValueType> SequentialRobinHoodHashTable::search(KeyType key) {
```
I think `std::optional` conveys intent better.

An argument against this would be that it actually has a space overhead (which is ridiculous, and a lot of C++ devs are also angry; they should really make this a language-level thing, not a STL thing).

Reply:
this is super cool didnt know u could do that